pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/minion/PinotTaskManager.java

```diff
@@ -0,0 +1,136 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.controller.helix.core.minion;
+
+import com.linkedin.pinot.common.config.AbstractTableConfig;
+import com.linkedin.pinot.common.config.PinotTaskConfig;
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import com.linkedin.pinot.controller.helix.core.minion.generator.PinotTaskGenerator;
+import com.linkedin.pinot.controller.helix.core.minion.generator.TaskGeneratorRegistry;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import org.apache.helix.task.TaskDriver;
+import org.apache.helix.task.WorkflowConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The class <code>PinotTaskManager</code> is the component inside Pinot Controller to periodically check the Pinot
+ * cluster status and schedule new tasks.
+ * <p><code>PinotTaskManager</code> is also responsible for checking the health status on each type of tasks, detect and
+ * fix issues accordingly.
+ */
+public class PinotTaskManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PinotTaskManager.class);
+
+  private final TaskDriver _taskDriver;
+  private final PinotHelixResourceManager _pinotHelixResourceManager;
+  private final PinotHelixTaskResourceManager _pinotHelixTaskResourceManager;
+  private final ClusterInfoProvider _clusterInfoProvider;
+  private final int _runFrequencyInSeconds;
+
+  private ScheduledExecutorService _executorService;
+
+  public PinotTaskManager(@Nonnull TaskDriver taskDriver, @Nonnull PinotHelixResourceManager pinotHelixResourceManager,
+      @Nonnull PinotHelixTaskResourceManager pinotHelixTaskResourceManager, int runFrequencyInSeconds) {
+    _taskDriver = taskDriver;
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    _pinotHelixTaskResourceManager = pinotHelixTaskResourceManager;
+    _clusterInfoProvider = new ClusterInfoProvider(pinotHelixResourceManager, pinotHelixTaskResourceManager);
+    _runFrequencyInSeconds = runFrequencyInSeconds;
+  }
+
+  /**
+   * Start the <code>PinotTaskManager</code>.
+   */
+  public void start() {
+    // Ensure job queue exists for all registered jobs
+    Map<String, WorkflowConfig> workflows = _taskDriver.getWorkflows();
+    for (String taskType : TaskGeneratorRegistry.getTaskGeneratorRegistry().keySet()) {
+      if (!workflows.containsKey(taskType)) {
+        _pinotHelixTaskResourceManager.createJobQueue(taskType);
+      }
+    }
+
+    // Schedule the execute method
```

Comment:
refactor this into a method that you can override in test classes, and then you should be able to call execute() directly